Third part of TDD + URM (#7)
sierisimo/publications: TDDWithURMAndKotlin/part2_increment.md

# TDD With URM and Kotlin - increment

> In the previous article we mentioned a way to start writing our implementation of URM using TDD. We also mentioned a lot of questions that came up while writing the `zero` function.

Reviewer: let's add the link to the previous article (it's never a bad idea!)

Now it's time to write the increment function.

## Increment function

> We know the `increment` function will be `I(x)`, where `x` represents a register in our _Unlimented Register Machine_. We are writing (for now) only the internal that will be used to represent our functionallity.

Reviewer: ```suggestion
We know that the `increment` function will be `I(x)`, where `x` represents a register in our _Unlimented Register Machine_. We are only writing the internal that will be used to represent our functionality -for now.
```

Reviewer: I think that @AnnaElizleal has a great point, `-` looks better when writing. Also, it's a good idea to keep it consistent across your articles 👍

> Following the approach we followed for the `zero` function, we start by writing a test on how we expect the function behaves with negative numbers:

Reviewer: ```suggestion
Following the same approach we used for the `zero` function, we'll start by writing a test on how we expect the function to behave with negative numbers:
```

Reviewer: Just getting rid of redundant terms and making the text a little more descriptiveness to the paragraph :)

```kotlin
@ParameterizedTest
@ValueSource(ints = [-2, -8, -20, -100])
fun `increment function throws exception with negative position`(position: Int) {
    assertThrows<IllegalArgumentException> { increment(registry, position) }

    assertNull(registry.getValueAtPosition(position))
}
```

> This initial test is in pro of the defensive model, we first need to be sure the function fails for some invalid cases. We could start with a different part of the code but this is the easiest to write because we already wrote this on the `zero` function:

Reviewer: ```suggestion
This initial test is in pro of the defensive model, but we first need to be sure the function fails for some invalid cases. We could start with a different part of the code but this is the easiest to write because we already wrote this on the `zero` function:
```

It's also a good idea to have a context for what you're talking about. You could even also add a: `but remember that we first need to...` to make people know that there's previous experiences from this, even if this is their gateway to your series~

```kotlin
fun increment(registry: Registry, position: Int){
    require(position > 0) { "Position must be positive number" }
}
```

> For our second test we need want to validate a "strict" rule: when a register has no value, it cannot be incremented. Our test should represent this and we should put a simple exception for it:

Reviewer: ```suggestion
For our second test we need want to validate a "strict" rule: when a register has no value, it cannot be incremented. Our test should represent this and we should place a simple exception for it:
```

Using synonyms also help with the pace when reading :)
(people might get confused if you use the same term all the time and are going back and forth over your writing)

```kotlin
@ParameterizedTest
@ValueSource(ints = [2, 4, 20, 40])
fun `increment function cannot work with empty register`(position: Int) {
    assertThrows<IllegalStateException> { increment(registry, position) }

    assertNull(registry.getValueAtPosition(position))
}
```

If we run our test at this point it will fail:

```
Expected java.lang.IllegalStateException to be thrown, but nothing was thrown.
org.opentest4j.AssertionFailedError: Expected java.lang.IllegalStateException to be thrown, but nothing was thrown.
```

> This exception tell us what we already know: our function is not throwing the right exception for empty registers. Let's fix that using the contract `checkNotNull` which is available on the kotlin stdlib (I'll write about contracts in another article):

Reviewer: ```suggestion
This exception tells us what we already know: our function is not throwing the right exception for empty registers. Let's fix that using the contract `checkNotNull` which is available via the kotlin stdlib (I'll write about contracts in another article):
```
Third person! And don't forget to revisit these `I'll write about contracts` later, so you can link in between your articles!

```kotlin
fun increment(registry: Registry, position: Int){
    require(position > 0) { "Position must be positive number" }
    checkNotNull(registry.getValueAtPosition(position)) { "Register must be initialized" }
}
```

Now our tests are passing again. Let's write the actual implementation.

## The real implementation

> The test for the real implementation will need a litle more setup. We know that the current approach will throw an expection if our test is just:

Reviewer: ```suggestion
The test for the real implementation will need a litle more setup. We know that the current approach will throw an exception if our test is just:
```
Typo~

```kotlin
@ParameterizedTest
@ValueSource(ints = [11, 15, 110, 200, Int.MAX_VALUE])
fun `increment function adds 1 to value of position`(position: Int) {
    increment(registry, position)

    assertEquals(1, registry.getValueAtPosition(position))
}
```

> This is because we wrote a test that will throw an exception when a registry does not contain a value and someone calls the `increment` function on it. But fortunately we already wrote tests for the `zero` function which means we can initialize our registry before calling increment:

Reviewer: ```suggestion
This is because we wrote a test that will throw an exception when a registry does not contain a value and someone calls the `increment` function on it. But fortunately we already wrote tests for the `zero` function, which means we can initialize our registry before calling `increment`:
```
Adding a comma and indicating increment refers to the function... to keep it consistent

```kotlin
@ParameterizedTest
@ValueSource(ints = [11, 15, 110, 200, Int.MAX_VALUE])
fun `increment function adds 1 to value of position`(position: Int) {
    zero(registry, position)
    increment(registry, position)

    assertEquals(1, registry.getValueAtPosition(position))
}
```

To fix the breaking test we need to modify our function:

```kotlin
fun increment(registry: Registry, position: Int) {
    require(position > 0) { "Position must be positive number" }

    val currentValue = registry.getValueAtPosition(position)
    checkNotNull(currentValue) { "Register must be initialized" }

    registry.setValueAtPosition(position, currentValue + 1)
}
```

> We store the value of `registry.getValueAtPosition(position)` and then call the contract `checkNotNull` (I'll write about contracts soon, I promise!)

Reviewer: You need to write about contracts soon!

> But testing with a single `1` is boring… Let's write something that operates over different values. To test dinamycally we will need to check a parameter, then call `increment` and get an expected value. We can simply call `assertEquals(parameter + 1, registry.getValueAtPosition(parameter))` but somehow it feels like cheating…

Reviewer: I think I've read this before 🤔

Author: 🤔

Using _parameterized tests_ from JUnit 5 we can create a single test that will cover different cases in a single method, and we will use `@CsvSource` to pass different parameters. `@CsvSource` takes a group of comma-separated-strings for every set of parameters and our method will ask for this parameters in the appropiate types:

```kotlin
@ParameterizedTest
@CsvSource("50,1,2", "50,5,6", "50,100,101")
fun `increment function adds 1 to any integer value`(
    position: Int,
    original: Int,
    expected: Int
){
    registry.setValueAtPosition(position, original)

    increment(registry, position)

    assertEquals(expected, registry.getValueAtPosition(position))
}
```

> This allow the test to be more dynamic and we can add more cases having more flexibility!

Reviewer: ```suggestion
This allows the test to be more dynamic and we can add more cases having more flexibility!
```
Third person :P

Let's write one test mixing what we learned:

```kotlin
@ParameterizedTest
@CsvSource("10,23,24", "15,99,100", "1000,1000,1001")
fun `check zero and increment can work together`(
    position: Int,
    original: Int,
    expected: Int
) {
    registry.setValueAtPosition(position, original)
    increment(registry, position)
    assertEquals(expected, registry.getValueAtPosition(position))

    zero(registry, position)
    assertEquals(0, registry.getValueAtPosition(position))

    increment(registry, position)
    assertEquals(1, registry.getValueAtPosition(position))
}
```

> Wow! That's a test. We start by setting manually a value on a position. Then we increment it to assert the increment works as expected (we know it does because we have a test to validate it!) and then we do something "new": we reset to 0 a register. We are validating something we didn't check before, that the `zero` function can reset any registry regardless of the value. Finally we `increment` and validate that `increment` works regardless of the value.

Reviewer: ```suggestion
Wow! That's a test. We start by setting manually a value on a position. Then we `increment` it to assert that `increment` works as expected (we know it does because we have a test to validate it!) and then we do something "new": we reset to 0 a register. We are validating something we didn't check before, that the `zero` function can reset any registry regardless of the value. Finally we `increment` and validate that `increment` works regardless of the value.
```
Consistency, muchacho!

Author: You right my amigo

---

## Conclusion

> We added a new function and also we validate our previous work is not affecting the new work we added. Everything was achieved through writing tests and fixing code!

Reviewer: ```suggestion
We added a new function and also validated that our previous work is not affected by the new code we added. Everything was achieved through writing tests and fixing code!
```
Tense consistency~

> There's still things that makes the tests feel like cheating… calling manually `registry.setValueAtPosition` and having a custom implementation for example. We will fix this in the next article where we will talk about mocks!

Reviewer: ```suggestion
There are still things that makes the tests feel like cheating… calling manually `registry.setValueAtPosition` and having a custom implementation for example. We will fix this in the next article where we will talk about mocks!
```
Don't forget to link your next article either!
